LINVAST.Imperative/Builders/Java/JavaASTBuilder.Types.cs

```diff
@@ -14,6 +14,193 @@ namespace LINVAST.Imperative.Builders.Java
+                var interfaceDecl = this.Visit(ctx.interfaceDeclaration()).As<TypeDeclNode>();
+                int declSpecsLine = ctxStartLine;
+                return new InterfaceNode(ctx.Start.Line, new DeclSpecsNode(declSpecsLine, modifiers, interfaceDecl.Identifier), interfaceDecl);
+            }
+            return new EmptyStatNode(ctx.Start.Line);
+        }
+
+        public override ASTNode VisitClassOrInterfaceModifier([NotNull] ClassOrInterfaceModifierContext ctx)
+        {
+
+            if (ctx.annotation() is { })
+                throw new NotImplementedException("annotations");
+
+            return new DeclSpecsNode(ctx.Start.Line, ctx.children.First().GetText());
+
+        }
+
+        public override ASTNode VisitTypeType([NotNull] TypeTypeContext ctx)
+        {
+
+            if (ctx.annotation() is { } && ctx.annotation().Any()) {
+                throw new NotImplementedException("annotations");
+            }
+
+            if (ctx.primitiveType() is { }) {
+                return this.Visit(ctx.primitiveType());
+
+            }
+
+            return this.Visit(ctx.classOrInterfaceType());
+
+        }
+
+
+        public override ASTNode VisitTypeList([NotNull] TypeListContext ctx)
+        {
+            var typeNameNodes = ctx.typeType().Select(c => this.Visit(c).As<TypeNameNode>()).ToArray<TypeNameNode>();
+
+
+            return new TypeNameListNode(ctx.Start.Line, typeNameNodes);
+        }
+
+        public override ASTNode VisitTypeParameters([NotNull] TypeParametersContext ctx)
+        {
+            return new TypeNameListNode(ctx.Start.Line, ctx.typeParameter().Select(c => this.Visit(c).As<TypeNameNode>()).ToArray());
+        }
+
+        public override ASTNode VisitTypeParameter([NotNull] TypeParameterContext ctx)
+        {
+            if (ctx.annotation() is { } && ctx.annotation().Any())
+                throw new NotImplementedException("annotations");
+
+            TypeNameListNode baseList;
+            if (ctx.typeBound() is { })
+                baseList = this.Visit(ctx.typeBound()).As<TypeNameListNode>();
+            else
+                baseList = new TypeNameListNode(ctx.Start.Line, new ArrayList<TypeNameNode>());
+
+            return new TypeNameNode(ctx.Start.Line, ctx.IDENTIFIER().GetText(), baseList.Types);
+        }
+
+        public override ASTNode VisitTypeBound([NotNull] TypeBoundContext ctx)
+        {
+            var typeNameNodes = ctx.typeType().Select(c => this.Visit(c).As<TypeNameNode>()).ToArray<TypeNameNode>();
+            return new TypeNameListNode(ctx.Start.Line, typeNameNodes);
+        }
+
+
+        public override ASTNode VisitPrimitiveType([NotNull] PrimitiveTypeContext ctx)
+        {
+
+            return new TypeNameNode(ctx.Start.Line, ctx.children.First().GetText());
+
+        }
+
+        public override ASTNode VisitClassType([NotNull] ClassTypeContext ctx)
+        {
+
+            if (ctx.annotation() is { } && ctx.annotation().Any()) {
```

Comment:
I guess just `if (ctx.annotation() is { })` is sufficient?

Reply:
No, if I remove ctx.annotation().Any() from condition tests regarding those functions, won't work. But _if(ctx.annotation().Any())_ is sufficient, because it is an array of annotations (according to rules). I removed every _ctx.annotation() is { }_ where there are together, except in function _VisitClassOrInterfaceModifier_ where there is one annotation.